EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
If you switch to range finder, the estimate will be the current height. It just needs the offset to be able to fuse it into the estimator (the difference between estimated z and sensor distance). So no matter what happened before, at the point you switch to the range finder, it will stay at the same height above the ground.